File: dart_troll/lib/src/dk/intro.dart
import 'package:dart_troll/src/core/audio.dart';
import 'package:dart_troll/src/core/scene.dart';
import 'package:dart_troll/src/core/util.dart';
import 'package:dart_troll/src/dk/height.dart';
import 'package:dart_troll/src/dk/sprites.dart';
import 'package:dart_troll/src/proto/animation.pb.dart';
import 'package:dart_troll/src/proto/primitives.pb.dart';
import 'package:dart_troll/src/proto/scene.pb.dart' as proto;

class IntroScene extends Scene {
  IntroScene() {
    scene = proto.Scene()
      ..id = 'intro'
      ..viewport = (Box()
        ..width = 640
        ..height = 480);
  }

  void setup() {
    final platformSpecs = [
      _PlatformSpec(3, [230, 97, -3]),
      _PlatformSpec(13, [80, 155, -3]),
      _PlatformSpec(13, [112, 210, -3]),
      _PlatformSpec(13, [80, 275, -3]),
      _PlatformSpec(13, [112, 335, -3]),
      _PlatformSpec(13, [80, 395, -3]),
      _PlatformSpec(14, [80, 455, -3]),
    ];

    final platforms = platformSpecs
        .asMap()
        .map((index, spec) {
          final platform = Platform(index, spec.size);
          platform.create(spec.position);
          return MapEntry(index, platform);
        })
        .values
        .toList();

    final ladderPositions = [
      [210, 72, -1],
      [242, 72, -1],
      [210, 112, -1],
      [242, 112, -1],
      [338, 112, -1],
      [305, 170, -1],
      [338, 170, -1],
      [305, 210, -1],
      [338, 210, -1],
      [305, 250, -1],
      [338, 250, -1],
      [305, 290, -1],
      [338, 290, -1],
      [305, 330, -1],
      [338, 330, -1],
      [305, 370, -1],
      [338, 370, -1],
      [305, 410, -1],
      [338, 410, -1],
    ];

    final ladders = ladderPositions.reversed.map<Ladder>((position) {
      return Ladder()..create(position);
    }).toList();

    final ladderFadingScript = AnimationScript()
      ..animation.add(Animation()..flash = (FlashAnimation()..repeat = 1));
    for (var i = 0; i < 4; ++i) {
      ladderFadingScript.animation[0].flash.delay = 700 * (i + 1);
      ladders[i * 2].playAnimationScript(ladderFadingScript);
      ladders[i * 2 + 1].playAnimationScript(ladderFadingScript);
    }

    final script = AnimationScript()
      ..id = 'intro'
      ..animation.addAll([
        Animation()..runScript = (RunScriptAnimation()..scriptId = 'dk_climb'),
        Animation()..timer = (TimerAnimation()..delay = 1000),
        Animation()
          ..runScript = (RunScriptAnimation()..scriptId = 'dk_landing'),
        Animation()..timer = (TimerAnimation()..delay = 300),
        Animation()
          ..runScript = (RunScriptAnimation()..scriptId = 'dk_jump')
          ..goTo = (GotoAnimation()
            ..destination = makeVector([130, 90])
            ..step = 1
            ..delay = 18)
          ..termination = Animation_TerminationCondition.ALL,
        Animation()..timer = (TimerAnimation()..delay = 300),
        Animation()..runScript = (RunScriptAnimation()..scriptId = 'dk_taunt'),
        Animation()..timer = (TimerAnimation()..delay = 1000),
      ]);

    final dk = DonkeyKong()
      ..create([300, 382, 0], frameIndex: 4)
      ..playAnimationScript(script, onDone: () {
        HeightScene()
          ..transition()
          ..setup();
      });

    dk.onScriptDone('dk_landing', () {
      Princess().create([270, 50, -1], frameIndex: 1);
      platforms[1].collapse();
      for (int i = 4; i < 7; ++i) {
        ladders[i * 2].destroy();
        ladders[i * 2 + 1].destroy();
      }
    });

    dk.onScriptDone('dk_jump', () {
      platforms[2].collapse();
      dk.playAnimationScriptById('dk_jump', onDone: () {
        platforms[3].collapse();
        dk.playAnimationScriptById('dk_jump', onDone: () {
          platforms[4].collapse();
          dk.playAnimationScriptById('dk_jump', onDone: () {
            platforms[5].collapse();
            dk.playAnimationScriptById('dk_jump', onDone: () {
              platforms[6].collapse();
            });
          });
        });
      });
    });

    playMusic('intro');
  }
}

class _PlatformSpec {
  int size;
  List<int> position;
  _PlatformSpec(this.size, this.position);
}

File: samples/donkey_kong/data/scripts/dk/intro.py
import dk.height
import dk.sprites
import proto.animation_pb2
import proto.key_binding_pb2
import pytroll.audio
import pytroll.events
import pytroll.input
import pytroll.scene


class IntroScene(pytroll.scene.PyTrollScene):
    def __init__(self):
        self.scene = proto.scene_pb2.Scene()
        self.scene.id = 'intro'
        self.scene.viewport.width = 640
        self.scene.viewport.height = 480

    def Setup(self):
        platform_sizes = [3, 13, 13, 13, 13, 13, 14]
        platforms = [dk.sprites.Platform(i, size)
                     for i, size in enumerate(platform_sizes)]
        platform_positions = [
            (262, 97, -3), (80, 155, -3), (112, 210, -3), (80, 275, -3),
            (112, 335, -3), (80, 395, -3), (80, 455, -3),
        ]
        for platform, pos in zip(platforms, platform_positions):
            platform.Create(pos)

        ladder_positions = [
            (210, 72, -1), (242, 72, -1),
            (210, 112, -1), (242, 112, -1),
            (338, 112, -1),
            (305, 170, -1), (338, 170, -1),
            (305, 210, -1), (338, 210, -1),
            (305, 250, -1), (338, 250, -1),
            (305, 290, -1), (338, 290, -1),
            (305, 330, -1), (338, 330, -1),
            (305, 370, -1), (338, 370, -1),
            (305, 410, -1), (338, 410, -1),
        ]
        ladders = [dk.sprites.Ladder(i).Create(pos, frame_index=0)
                   for i, pos in enumerate(ladder_positions)]

        donkey_kong = dk.sprites.DonkeyKong()
        donkey_kong.Create((300, 382, 0), frame_index=4)
        self.__Cutscene(donkey_kong, platforms, ladders)

    def __Cutscene(self, donkey_kong, platforms, ladders):
        script = proto.animation_pb2.AnimationScript()
        script.id = 'intro'
        script.animation.add().run_script.script_id = 'dk_climb'
        script.animation.add().timer.delay = 1000
        script.animation.add().run_script.script_id = 'dk_landing'
        script.animation.add().timer.delay = 300

        destroy_platforms = script.animation.add()
        destroy_platforms.termination = proto.animation_pb2.Animation.ALL
        goto = destroy_platforms.go_to
        goto.destination.x, goto.destination.y = 130, 90
        goto.step = 1
        goto.delay = 18
        destroy_platforms.run_script.script_id = 'dk_jump'

        script.animation.add().timer.delay = 300
        script.animation.add().run_script.script_id = 'dk_taunt'
        script.animation.add().timer.delay = 1000
        donkey_kong.PlayAnimationScript(
            script, lambda e: self.Transition(dk.height.HeightScene()))
        pytroll.audio.PlayMusic('intro')

        def LandingHandler():
            dk.sprites.Princess().Create((270, 50, -1), frame_index=1)
            platforms[1].Collapse()
            for i in range(4, 7):
                ladders[-(i * 2 + 1)].Destroy()
                ladders[-(i * 2 + 2)].Destroy()
        pytroll.events.OnEvent(
            pytroll.events.AnimationScriptDone(donkey_kong.id, 'dk_landing'),
            lambda e: LandingHandler())

        def DestroyPlatform(index):
            platforms[index].Collapse()
            if index < 6:
                donkey_kong.PlayAnimationById(
                    'dk_jump', lambda e: DestroyPlatform(index + 1))
        pytroll.events.OnEvent(
            pytroll.events.AnimationScriptDone(donkey_kong.id, 'dk_jump'),
            lambda e: DestroyPlatform(2))

        script = proto.animation_pb2.AnimationScript()
        script.animation.add().flash.repeat = 1
        for i in range(4):
            script.animation[0].flash.delay = 700 * (i + 1)
            ladders[-(i * 2 + 1)].PlayAnimationScript(script)
            ladders[-(i * 2 + 2)].PlayAnimationScript(script)

    def HandleInput(self, input_event):
        if input_event.HasField('key_event'):
            pressed = proto.key_binding_pb2.Trigger.PRESSED
            if input_event.key_event.key == 'SPACE' and \
                    input_event.key_event.key_state == pressed:
                self.Transition(dk.height.HeightScene())
